vernac/vernacentries.ml

```diff
@@ -1069,31 +1068,53 @@ let add_subnames_of ns full_n n =
         let n_elim = Id.of_string (Id.to_string mip.mind_typename ^ s) in
         match Nametab.extended_global_of_path (Libnames.make_path dp n_elim) with
         | exception Not_found -> ns
-        | n_elim -> NSet.add n_elim ns)
+        | n_elim -> (len, n_elim) :: ns)
       ns Sorts.all_families
 
-let interp_filter_in m = function
-  | ImportAll ->  Libobject.Unfiltered
-  | ImportNames ns ->
-    let module NSet = Globnames.ExtRefSet in
-    let dp_m = Nametab.dirpath_of_module m in
-    let ns =
-      List.fold_left (fun ns (n,etc) ->
-          let full_n =
-            let dp_n,n = repr_qualid n in
-            make_path (append_dirpath dp_m dp_n) n
-          in
-          let n = try Nametab.extended_global_of_path full_n
-            with Not_found ->
-              CErrors.user_err
-                Pp.(str "Cannot find name " ++ pr_qualid n ++ spc() ++
-                    str "in module " ++ pr_qualid (Nametab.shortest_qualid_of_module m))
-          in
-          let ns = NSet.add n ns in
-          if etc then add_subnames_of ns full_n n else ns)
-        NSet.empty ns
-    in
-    Libobject.Names ns
+let interp_names m ns =
+  let dp_m = Nametab.dirpath_of_module m in
+  let ns =
+    List.fold_left (fun ns (n,etc) ->
+        let len, full_n =
+          let dp_n,n = repr_qualid n in
+          List.length (DirPath.repr dp_n), make_path (append_dirpath dp_m dp_n) n
+        in
+        let n = try Nametab.extended_global_of_path full_n
+          with Not_found ->
+            CErrors.user_err
+              Pp.(str "Cannot find name " ++ pr_qualid n ++ spc() ++
+                  str "in module " ++ pr_qualid (Nametab.shortest_qualid_of_module m))
+        in
+        (* TODO dumpglob? *)
+        let ns = (len,n) :: ns in
+        if etc then add_subnames_of len ns full_n n else ns)
+      [] ns
+  in
+  ns
+
+let cache_name (len,n) =
+  let open Globnames in
+  let open GlobRef in
+  match n with
+  | SynDef kn -> Syntax_def.import_syntax_constant (len+1) (Nametab.path_of_syndef kn) kn
+  | TrueGlobal (VarRef _) -> assert false
+  | TrueGlobal gr ->
+    (* imports even local constants, do we care? *)
+    Nametab.(push (Exactly (len+1)) (path_of_global gr) gr)
+
+let cache_names (_,ns) = List.iter cache_name ns
+
+let subst_names (subst,ns) = List.Smart.map (on_snd (Globnames.subst_extended_reference subst)) ns
+
+let inExportNames = Libobject.declare_object
+    (Libobject.global_object "EXPORTNAMES"
+       ~cache:cache_names ~subst:(Some subst_names)
```

Comment:
It's possible to have something like
~~~coq
Module F(A:T).
Export A(x).
End F.
Module Import M := F X.
Check x.
~~~
This seems kinda strange but works OK.
Probably should be tested though.